runners/direct-java/src/main/java/org/apache/beam/runners/direct/TestStreamEvaluatorFactory.java

```diff
@@ -130,8 +155,9 @@ public TransformResult finishBundle() throws Exception {
             .advance(((ProcessingTimeEvent<T>) event).getProcessingTimeAdvance());
       }
       index++;
-      checkState(inUse.compareAndSet(true, false),
-          "The InUse flag of a %s was changed while the source evaluator was executing. "
+      checkState(
```

Comment:
I think this was part of what I was missing.

I'd suggest passing in the `TestStreamEvaluatorFactory` here, and adding a method like "returnEvaluator", so that the interactions with the evaluators map are fully contained in the `TestStreamEvaluatorFactory`